### PR TITLE
Catch invalid broadcasts in pointwise reduce fusion

### PR DESCRIPTION
**Pull request by shivadbhavsar**

### CODE REVIEW

File: src/fuse_reduce.cpp [hunk 1]
```diff
@@ -197,13 +197,31 @@ struct find_pointwise_reduce
         return match::name("fused_reduce")(match_broadcastable_input("pointwise", "pointwise"));
     }
 
+    bool is_valid_broadcast(const instruction_ref b, const std::vector<size_t> reduce_axes) const
+    {
+        std::vector<size_t> broadcast_indices;
+        auto bstrides = b->get_shape().strides();
+
+        return std::all_of(
+            reduce_axes.begin(), reduce_axes.end(), [&](auto a) { return bstrides.at(a) == 0; });
```
Comment (pfultz2) [inline src/fuse_reduce.cpp:206]: I think you need to check if the non-reduce axes are not being broadcasted as well.

File: src/fuse_reduce.cpp [hunk 2]
```diff
@@ -197,13 +197,31 @@ struct find_pointwise_reduce
...
         auto input         = r.instructions["pointwise"];
         const auto* pm     = input->module_inputs().front();
         const auto* old_rm = reduce->module_inputs().front();
 
+        if(contains(r.instructions, "broadcast"))
+        {
+            auto axes       = reduce->get_operator().to_value().at("axes").to_vector<size_t>();
+            auto broadcast  = r.instructions["broadcast"];
+            auto fbroadcast = r.instructions["final_broadcast"];
+            if(not(is_valid_broadcast(broadcast, axes) and is_valid_broadcast(fbroadcast, axes)))
```
Comment (pfultz2) [inline src/fuse_reduce.cpp:221]: You should just check "broadcast" and not "final_broadcast"(which is the broadcast after contiguous if there is one).

File: src/fuse_reduce.cpp [hunk 3]
```diff
@@ -389,8 +411,8 @@ void fuse_reduce::apply(module_pass_manager& mpm) const
     {
         if(enable_rewrite_reshapes)
             mpm.run_pass(rewrite_reshapes<reduce_reshape>{});
-        match::find_matches(
-            mpm, find_reduce_pointwise{}, find_pointwise_reduce{}, find_reduce_reduce{});
+        match::find_matches(mpm, find_reduce_pointwise{}, find_pointwise_reduce{});
+        match::find_matches(mpm, find_reduce_reduce{});
```
Comment (pfultz2) [inline src/fuse_reduce.cpp:415]: Why is this moved out?

Comment (shivadbhavsar) [reply src/fuse_reduce.cpp:415]: because when it hits the broadcast condition, its not fusing the other fused_reduce in the input

Comment (pfultz2) [reply src/fuse_reduce.cpp:415]: We might need to add this to the matcher then.

File: src/fuse_reduce.cpp [hunk 4]
```diff
@@ -197,13 +197,35 @@ struct find_pointwise_reduce
...
         const auto* pm     = input->module_inputs().front();
         const auto* old_rm = reduce->module_inputs().front();
 
+        if(contains(r.instructions, "broadcast"))
+        {
+            auto axes      = reduce->get_operator().to_value().at("axes").to_vector<size_t>();
+            auto broadcast = r.instructions["broadcast"];
+            if(not is_valid_broadcast(broadcast, axes))
+                return;
+        }
```
Comment (pfultz2) [inline src/fuse_reduce.cpp:227]: This check needs to be applied to all matchers.

**Comment by pfultz2** [comment]
So we could create a matcher like this to check the axes:

```cpp
template<class M>
static auto match_broadcast_axes(M m)
{
    return make_basic_fun_matcher([=](matcher_context& ctx, instruction_ref ins) {
        optional<instruction_ref> result = m.match(ctx, ins);
        if(contains(ctx.instructions, "broadcast"))
        {
            auto axes      = ins->get_operator().to_value().at("axes").to_vector<size_t>();
            auto broadcast = r.instructions["broadcast"];
            if(not is_valid_broadcast(broadcast, axes))
                return nullopt;
        }
        return result;
    });
}
```

And then we could change `match_broadcastable_input` to use it like this:

```cpp
static auto match_broadcastable_input(const std::string& op, const std::string& name)
{
    auto match_op                 = match::name(op)(used_once_except_broadcast()).bind(name);
    auto match_op_input           = any_input(match_op, match::used_once());
    auto broadcast_match_op_input = any_input(match_broadcast(match_op), match::used_once());
    return match::any_of(match_op_input, match_broadcast_axes(broadcast_match_op_input));
}
```

I dont know if there is a better way of doing this.

**Review by pfultz2 — APPROVED** [review]

**Comment by pfultz2** [comment]
Maybe a verify test could be added for this as well.

**Comment by shivadbhavsar** [comment]
> Maybe a verify test could be added for this as well.

how can I force the verify test to run with `MIGRAPHX_DISABLE_LAYERNORM_FUSION=1`? or will CI do that at some point?

**Comment by pfultz2** [comment]
> how can I force the verify test to run with MIGRAPHX_DISABLE_LAYERNORM_FUSION=1? or will CI do that at some point?

Well I am hoping to remove layernorm fusion soon in #3465. For now the verify tests can run with the layernorm fusion, and when it gets removed in #3465, this test will check that it works correctly.

File: src/fuse_reduce.cpp [hunk 5]
```diff
@@ -174,12 +174,53 @@ static auto any_input(Ms... ms)
     return match::any_of[match::inputs()](match::any(ms...).bind("input"));
 }
 
+bool is_valid_broadcast(const instruction_ref b, const std::vector<size_t>& reduce_axes)
+{
+    std::vector<size_t> broadcast_axes;
+    auto bstrides = b->get_shape().strides();
+
+    for(size_t i = 0; i < bstrides.size(); ++i)
```
Comment (TedThemistokleous) [inline src/fuse_reduce.cpp:182]: Is the intent here to skip the first index?, this starts at 1
